torch/lib/c10d/reducer.cpp

```diff
@@ -563,9 +559,31 @@ void Reducer::mark_variable_ready(VariableIndex index) {
     if (find_unused_parameters_) {
       // H2D from local_used_maps_ to local_used_maps_dev_
       for (size_t i = 0; i < local_used_maps_.size(); i++) {
-        // We do async H2D to avoid the blocking overhead. The async copy and
-        // allreduce respect the current stream, so will be sequenced correctly.
-        local_used_maps_dev_[i].copy_(local_used_maps_[i], true);
+        if (local_used_maps_dev_[i].is_cuda()) {
+          // Note [local_used_maps_ -> local_used_maps_dev copying]
+          // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+          // We do async H2D to avoid the blocking overhead. The async copy and
+          // allreduce respect the current stream, so will be sequenced correctly.
+          //
+          // Correct sequencing with respect to host operations is also essential.
+          // The H2D copy_ is stream ordered, while the host's changes to local_used_maps_
+          // are host ordered. If a large backlog of cuda-stream work pushes the copy_ far
+          // into the future, and if no blocking calls occur between now and finalize_backward()**
+          // such that finalize_backward() re-zeroes local_used_maps_ on the host
+          // before the stream executes the copy_, copy_ will read those zeros instead of the
+          // values we thought we told it to read here.
+          // Copying local_used_maps_[i] to a pinned temporary (which the pinned caching
+          // allocator should supply asynchronously) avoids this nasty, rare race condition.
+          //
+          // ** In the hoped-for case where all params are used, DDP itself won't do any
+          // blocking work between now and the re-zeroing, so the danger is real.
+          auto local_used_maps_tmp = local_used_maps_[i].pin_memory();
+          // Defensively ensures a deep copy to a pinned temporary
+          TORCH_INTERNAL_ASSERT(local_used_maps_tmp.data_ptr() != local_used_maps_[i].data_ptr())
```

Comment:
Curious, does `pin_memory` always guarantee that the resulting tensor has a different data_ptr than the unpinned tensor? The docs for `pin_memory` (https://pytorch.org/docs/stable/notes/cuda.html?highlight=pin_memory) say that it "returns a copy of the object, with data put in a pinned region" but is it possible that the address was previously in a pinned region so data_ptr() doesn't change?

Reply:
In theory yes, but pinned memory is managed by its own caching allocator, so I'd expect addresses of unpinned cpu allocations and pinned cpu allocations to be distinct, unless the pinned allocator is somehow leaking pinned memory (pinning it then forgetting about it). Such a leak would be bad because pinned memory is not a resource you should use excessively (pinning memory prevents the OS from swapping their physical pages to disk as needed, and can completely kill the system if it leaks too much).